### PR TITLE
 Add Basic Authentication deployment files

### PR DESCRIPTION
**Pull request by maelk**
Support the full matrix of cases for :
- Keepalived enabled/disabled
- TLS enabled/disabled
- basic authentication enabled/disabled

Please refer to docs/deploying.md for additional information

This PR is based on #631 .

### CODE REVIEW

File: ironic-deployment/basic-auth/ironic-rpc-auth-config-tpl [hunk 1]
```diff
@@ -0,0 +1,4 @@
+[json-rpc]
+auth_strategy=basic
+username=${IRONIC_USERNAME}
+password=${IRONIC_PASSWORD}
```
Comment (zaneb) [inline ironic-deployment/basic-auth/ironic-rpc-auth-config-tpl:4]: Note that prior to https://review.opendev.org/742474 the names here were `http_basic_username` and `http_basic_password`. You might want to pass both for a while.

File: docs/deploying.md [hunk 2]
```diff
@@ -130,13 +130,73 @@ Then you need to copy all your certificates and build the kustomization for
...
+  cat $BMOPATH/ironic-deployment/basic-auth/ironic-auth-config-tpl | envsubst > \
+  "${KSTM_PATH}/ironic-auth-config"
+
+  cat $BMOPATH/ironic-deployment/basic-auth/ironic-inspector-auth-config-tpl | \
+  envsubst > "${KSTM_PATH}/ironic-inspector-auth-config"
+
+  cat $BMOPATH/ironic-deployment/basic-auth/ironic-rpc-auth-config-tpl | \
+  envsubst > "${KSTM_PATH}/ironic-rpc-auth-config"
+
+  htpasswd -c -b -B "${KSTM_PATH}/HTTP_BASIC_HTPASSWD" abc def
```
Comment (zaneb) [inline docs/deploying.md:195]: `abc def` should be `${IRONIC_USERNAME} ${IRONIC_PASSWORD}`

File: docs/deploying.md [hunk 3]
```diff
@@ -130,13 +130,73 @@ Then you need to copy all your certificates and build the kustomization for
...
+### Deploying with Basic Authentication
+
+You can deploy each of the alternatives above with basic authentication
+enabled. For this, you first need to follow the instructions of the above setup
+other than running kustomize.
+
+Then define your username and password
+
+```sh
+  export IRONIC_USERNAME="<username>"
```
Comment (zaneb) [inline docs/deploying.md:149]: Nobody cares what the username is, so you could just pick one.

File: docs/deploying.md [hunk 4]
```diff
@@ -130,13 +130,73 @@ Then you need to copy all your certificates and build the kustomization for
...
+   cp <path-to-tls-cert-key> $BMOPATH/ironic-deployment/tls/keepalived/tls.key
+   ./tools/bin/kustomize build $BMOPATH/ironic-deployment/tls/keepalived | \
    kubectl apply -f -
 ```
 
+### Deploying with Basic Authentication
+
+You can deploy each of the alternatives above with basic authentication
+enabled. For this, you first need to follow the instructions of the above setup
+other than running kustomize.
```
Comment (zaneb) [inline docs/deploying.md:144]: This looks pretty complicated. You could wrap it all up in a script (including generating a random password).

**Review by zaneb — COMMENTED** [review]
This looks basically good, but I left come comments on subtleties inline.
FYI my implementation of this for OpenShift is in openshift/machine-api-operator#657 if you want to check anything there.

**Comment by maelk** [comment]
@zaneb I have address all your comments in the new script (tools/deploy.sh). Thank you for the input.

/test-integration

**Comment by dhellmann** [comment]
I would like to freeze go code changes for a few days to try to land https://github.com/metal3-io/baremetal-operator/pull/650 without having to rebase it, because rebasing will mean redoing the work from scratch.

/hold

**Comment by dhellmann** [comment]
#655 has merged

/hold cancel

**Pull request merged**